runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.beam.sdk.testing.UsesCrossLanguageTransforms;
+import org.apache.beam.sdk.testing.ValidatesRunner;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Filter;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.sdk.values.TypeDescriptors;
+import org.apache.beam.vendor.grpc.v1p13p1.io.grpc.ConnectivityState;
+import org.apache.beam.vendor.grpc.v1p13p1.io.grpc.ManagedChannel;
+import org.apache.beam.vendor.grpc.v1p13p1.io.grpc.ManagedChannelBuilder;
+import org.apache.beam.vendor.grpc.v1p13p1.io.grpc.Server;
+import org.apache.beam.vendor.grpc.v1p13p1.io.grpc.ServerBuilder;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableMap;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Test External transforms. */
+@RunWith(JUnit4.class)
+public class ExternalTest implements Serializable {
+  @Rule public transient TestPipeline testPipeline = TestPipeline.create();
+
+  private static final String TEST_URN_SIMPLE = "simple";
+  private static final String TEST_URN_LE = "le";
+  private static final String TEST_URN_MULTI = "multi";
+
+  private static String pythonServerCommand;
+  private static Integer expansionPort;
+  private static String localExpansionAddr;
+  private static Server localExpansionServer;
+
+  @BeforeClass
+  public static void setUp() throws IOException {
+    pythonServerCommand = System.getProperty("pythonTestExpansionCommand");
+    expansionPort = Integer.valueOf(System.getProperty("expansionPort"));
+    int localExpansionPort = expansionPort + 100;
+    localExpansionAddr = String.format("localhost:%s", localExpansionPort);
+
+    localExpansionServer =
+        ServerBuilder.forPort(localExpansionPort).addService(new ExpansionService()).build();
+    localExpansionServer.start();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    localExpansionServer.shutdownNow();
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+  public void expandSingleTest() {
+    PCollection<Integer> col =
+        testPipeline
+            .apply(Create.of(1, 2, 3))
+            .apply(External.of(TEST_URN_SIMPLE, new byte[] {}, localExpansionAddr));
+    PAssert.that(col).containsInAnyOrder(2, 3, 4);
+    testPipeline.run();
+  }
+
+  @Test
+  @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+  public void expandMultipleTest() {
```

Comment:
Do we have a test that uses Python transforms from Java ? These tests seems to be using Java transforms defined below, right ?

Reply:
This test uses the Python expansion service:
https://github.com/apache/beam/pull/7954/files/e8f99c6ff8d5f485f39b6df0f92a7e81681cca62#diff-889f2561e62c0f998c24e4a4079cff78R137